src/backend/vulkan/src/window.rs

```diff
@@ -34,69 +38,124 @@ impl Drop for RawSurface {
-            .map(|&extension| {
-                match extension {
-                    #[cfg(unix)]
-                    vk::VK_KHR_XLIB_SURFACE_EXTENSION_NAME => {
-                        use winit::os::unix::WindowExt;
-
-                        let xlib_loader = match ext::XlibSurface::new(entry, &self.raw.0) {
-                            Ok(loader) => loader,
-                            Err(e) => {
-                                error!("XlibSurface failed: {:?}", e);
-                                return None;
-                            }
-                        };
-
-                        let info = vk::XlibSurfaceCreateInfoKHR {
-                            s_type: vk::StructureType::XlibSurfaceCreateInfoKhr,
-                            p_next: ptr::null(),
-                            flags: vk::XlibSurfaceCreateFlagsKHR::empty(),
-                            window: window.get_xlib_window().unwrap() as *const _,
-                            dpy: window.get_xlib_display().unwrap() as *mut _,
-                        };
-
-                        unsafe {
-                            xlib_loader.create_xlib_surface_khr(&info, None).ok()
-                        }
-                    }
-                    #[cfg(windows)]
-                    vk::VK_KHR_WIN32_SURFACE_EXTENSION_NAME => {
-                        use kernel32;
-                        use winit::os::windows::WindowExt;
-
-                        let win32_loader = ext::Win32Surface::new(entry, &self.raw.0)
-                            .expect("Unable to load win32 surface functions");
-
-                        unsafe {
-                            let info = vk::Win32SurfaceCreateInfoKHR {
-                                s_type: vk::StructureType::Win32SurfaceCreateInfoKhr,
-                                p_next: ptr::null(),
-                                flags: vk::Win32SurfaceCreateFlagsKHR::empty(),
-                                hinstance: kernel32::GetModuleHandleW(ptr::null()) as *mut _,
-                                hwnd: window.get_hwnd() as *mut _,
-                            };
-
-                            win32_loader.create_win32_surface_khr(&info, None).ok()
-                        }
-                    }
-                    // TODO: other platforms
-                    _ => None,
-                }
-            })
-            .find(|x| x.is_some())
-            .expect("Unable to find a surface implementation.")
-            .unwrap();
+        let surface = {
+            use std::mem::transmute;
+
+            let xlib_loader = ext::XlibSurface::new(entry, &self.raw.0)
+                                .expect("XlibSurface failed");
+
+            let info = vk::XlibSurfaceCreateInfoKHR {
+                s_type: vk::StructureType::XlibSurfaceCreateInfoKhr,
+                p_next: ptr::null(),
+                flags: vk::XlibSurfaceCreateFlagsKHR::empty(),
+                window: unsafe { transmute(window) },
```

Comment:
is unsafe transmute really needed here as opposed to a simple `window as _`?

Reply:
`Window` type in Xlib header file is defined as `XID` (and therefore `unsigned long` in LP64 architecture), a pointer-sized integer type, not pointer. Thus `Window` should be `usize`. However, many crates including `x11` just define `Window` as `*mut c_void`, so `transmute` is necessary.

Reply:
Maybe I should open an issue ticket in `x11` repository

Reply:
Besides, we should cast `Window` as `c_ulong` to pass to `XGetGeometry ` and `XGetWindowAttributes` (due to `drawable` parameter is defined as `c_ulong` (aka `XID`)

Reply:
> However, many crates including x11 just define Window as *mut c_void, so transmute is necessary.

I'm not sure I follow. Conversions from `*mut _` to `usize` don't require `transmute`.

Reply:
OK, I corrected conversions, getting rid of `transmute`. Sorry for confusion.